javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -127,7 +127,7 @@ private Lazy(Supplier<? extends T> supplier) {
 
     public Option<T> filter(Predicate<? super T> predicate) {
         T v = get();
-        return predicate.test(v) ? Option.some(v) : Option.none();
+        return Option.when(predicate.test(v), () -> v);
```

Comment:
Ohh sh... - the implementation of Option.when is broken: if the supplier returns a null we create currently no Some instance, see https://github.com/javaslang/javaslang/blob/master/javaslang/src/main/java/javaslang/control/Option.java#L137

Option.when needs to be fixed to call some(...) instead of of(...)